Refactored how the middleware is configured, added depth calculation and throttling and added Rertry-After header on blocked requests, as well as error handling

evanmcneely opened the pull request:
### Summary
1. Refactored how the middle ware is configured, reducing 3 config parameters into 1. There is now only 2 arguments that need to be passed into the middelware function.
2. Added rate limiting data to `res.locals` for compatibility with the logger
3. confirming `req.ips[0] || req.ip` is the best way to access the ip-address for the express in express.
4. Added a retry after header to a blocked request that works with token bucket and sliding window log/counter
5. refactored the complexity analysis to calculate depth
6. Added tests for middleware configuration, retry-after calculations, and depth calculation
7. Added testing for @includes and @skip directives. We can ignore these for now as they are non-critical, only improve accuracy of analysis
8. Added error handling to middleware.

### Type of Change
- [x] New feature (non-breaking change which adds functionality)
- [x] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [x] This change requires a documentation update

### Issues
closes #74 
closes #47 
closes #41
closes #37 
closes #83 
closes #39 
closes #73 
closes #89 
closes #25 
closes #27 

### Evidence
![Screen Shot 2022-08-02 at 7 23 44 AM](https://user-images.githubusercontent.com/89324687/182408773-6e725f5f-dfdc-406e-b327-e185f57c3e51.png)

shalarewicz commented on src/middleware/index.ts:
```diff
@@ -1,47 +1,61 @@
-import { TypeWeightConfig } from '../@types/buildTypeWeights';
+import { ExpressMiddlewareConfig, ExpressMiddlewareSet } from '../@types/expressMiddleware';
 
 // FIXME: Will the developer be responsible for first parsing the schema from a file?
```
I think these `FIXME:`s can be removed.

shalarewicz commented on src/middleware/index.ts:
```diff
@@ -1,47 +1,61 @@
- * Defaults to {mutation: 10, object: 1, field: 0, connection: 2}
+ * @param {ExpressMiddlewareConfig} middlewareConfig
+ *  - One parameter to configure redis client, typeWeights and rate limiting parameters
+ *      - Ratelimiter is required in the setup of the middleware. Developers must explicitly specify this
```
These comments may appear in the docstring / hover text when a developer is using the package. I think they can be cleaned up a little and be a little more concise. 

`RateLimiter` should be `rateLimiter`
`ioredis` should be `redis`

shalarewicz commented on src/middleware/index.ts:
```diff
@@ -80,28 +85,36 @@ export function expressRateLimiter(
+            if (!rateLimiterResponse.success && !middlewareSetup.dark) {
+                // calculate the time the client should wait to send anouther query by comparing
+                // the differnce between tokens and complexity and multipying by the refill rate
+                const timeToWaitInMs =
```
Good idea, but this calculation is only applicable to `TOKEN_BUCKET`. I think we should add this to `RateLimiterResponse` since it will be much easier for the calculation to occur when we have access to all of the information. 
- `FIXED_WINDOW` - would need start of last window
- `SLIDING_WINDOW_LOG` - probably easiest for this computation to be done while interacting with redis. 
- `SLIDING_WINDOW_COUNTER` - We should be able to solve for this with pure math but again the rate limiter could return this. 
- `LEAKY_BUCKET` - similar to what you're doing here. 

This number should be rounded up to the nearest millisecond.

evanmcneely replied:
Ok, I'll probably leave this as a `fixme` and create an issue to calculate this once all the rate limit algorithms are done. Unless we want to task everyone with adding this to their current implementation?

shalarewicz replied:
I'm fine with tackling it later since it will involve updating tests. Maybe just add a condition to check if `TOKEN_BUCKET` is being used.

shalarewicz reviewed (changes requested):
Good work! This is much cleaner. My main comment is on the retry header computations.

shalarewicz commented:
Can you add a `keyExpiry` parameter to the redis options? This can then be passed down to each rate limiter

evanmcneely commented:
> Can you add a `keyExpiry` parameter to the redis options? This can then be passed down to each rate limiter

I've added this and addressed other comments. I'm going to create tests for all these parameters now.

shalarewicz commented on src/@types/expressMiddleware.ts:
```diff
@@ -0,0 +1,31 @@
+import { RedisOptions } from 'ioredis';
```
This file should be renamed `expressMiddleware.d.ts`

evanmcneely replied:
I have done this. I won't push changes until a few PR's close because I've already taken the time to pull them and work them and make them compatible.

evanmcneely commented:
closes #89 89

shalarewicz commented on src/rateLimiters/slidingWindowCounter.ts:
```diff
@@ -153,18 +155,54 @@ class SlidingWindowCounter implements RateLimiter {
+        // Size and proportion of the window in seconds
+        const windowSizeSeconds = this.windowSize / 1000;
+        const rollingWindowProportionSeconds = windowSizeSeconds * rollingWindowProportion;
+        // Tokens available for the request to use
```
Great documentation!

shalarewicz reviewed (approved):
This is a huge quality of life update! Great work!

shalarewicz commented on src/analysis/buildTypeWeights.ts:
```diff
@@ -49,20 +48,20 @@ export const defaultTypeWeightsConfig: TypeWeightSet = {
  * @param {(GraphQLObjectType | GraphQLInterfaceType)} type
  * @param {TypeWeightObject} typeWeightObject
  * @param {TypeWeightSet} typeWeights
+ * @param enforceBoundedLists
```
add `{boolean}` here

shalarewicz commented on src/analysis/buildTypeWeights.ts:
```diff
@@ -142,28 +169,28 @@ function parseObjectFields(
-                                throw new Error(
-                                    `ERROR: buildTypeWeights: Unbouned list complexity not supported. Query results should be limited with ${KEYWORDS}`
-                                );
+                                // if there is no argument or default value, return 0 complexity
```
The comment says return 0 but the code could returning something else

evanmcneely replied:
This has been updated with a more descriptive comment

shalarewicz replied:
@evanmcneely I don't see this update

shalarewicz commented on src/analysis/buildTypeWeights.ts:
```diff
@@ -372,28 +409,36 @@ function parseTypes(schema: GraphQLSchema, typeWeights: TypeWeightSet): TypeWeig
 ): TypeWeightObject {
-    if (!schema) throw new Error('Missing Argument: schema is required');
+    try {
+        if (!schema) throw new Error('Missing Argument: schema is required');
```
This error will get caught be the try catch block and the primary error will be `Error in expressGraphQLRateLimiter when parsing schema object: ${err}`. Is that intended?

evanmcneely replied:
I think it was yes

shalarewicz commented on src/utils/redis.ts:
```diff
@@ -8,10 +8,13 @@ const clients: Redis[] = [];
+        const client: Redis = new Redis(options); // Default port is 6379 automatically
+        clients.push(client);
+        return client;
+    } catch (err) {
```
I don't think this will have any effect. the redis client is an event emitter and will emit an `error` event if there is a connection issue. I believe it only tries to connect then it tries to perform an operation on the redis store.

evanmcneely replied:
Ok, I've removed this try-catch

shalarewicz commented on test/analysis/buildTypeWeights.test.ts:
```diff
@@ -1176,6 +1258,12 @@ describe('Test buildTypeWeightsFromSchema function', () => {
             );
         });
 
+        test('there is an unbounded list and user chooses to enforce bounded lists', () => {
```
Do we have a test where there is an unbounded list and the user chose not to enforce unbounded lists?

evanmcneely replied:
I have accounted for that in the weightFunction but there is no test. I will make one.

evanmcneely replied:
Although every other test implicitly checking for this.

shalarewicz commented on test/analysis/typeComplexityAnalysis.test.ts:
```diff
@@ -890,6 +854,170 @@ describe('Test getQueryTypeComplexity function', () => {
+                query = `
+                    query {
+                        hero(episode: EMPIRE) {
+                            ...@include(if: false) {
```
I would expect this function wouldn't even be called. Could be worth adding an assertion for this to avoid unnecessary work.

shalarewicz commented on test/middleware/express.test.ts:
```diff
@@ -34,20 +34,20 @@ const schema: GraphQLSchema = buildSchema(`
                     id: ID!
                     name: String!
-                    friends: [Character]
+                    friends: [Character] @listCost(cost: 10)
```
I think its worth using differing lists costs on line 37, 44, 50 to ensure the correct cost is used?

shalarewicz replied:
Actually, it doesn't like these are used in these tests

evanmcneely replied:
I'm not sure what you mean by this.

shalarewicz reviewed (approved):
Error handling looks great and will be very helpful moving forward

evanmcneely commented:
![Screen Shot 2022-08-02 at 7 23 44 AM](https://user-images.githubusercontent.com/89324687/182494204-68c16980-372c-4737-8198-8f69344ee777.png)

shalarewicz commented on src/analysis/ASTParser.ts:
```diff
@@ -139,8 +141,45 @@ class ASTParser {
+         * 2. there is a directive named inlcude and the value is true
+         * 3. there is a directive named skip and the value is false
+         */
+        // const directive = node.directives;
```
Opened #95 to track this

donjewey commented on src/analysis/buildTypeWeights.ts:
```diff
@@ -113,20 +112,28 @@ function parseObjectFields(
+                            if (
+                                dir.arguments &&
+                                dir.arguments[0].value.kind === Kind.INT &&
+                                Number(dir.arguments[0].value.value) >= 0
```
good thinking on another improper configuration check

donjewey commented on src/middleware/index.ts:
```diff
@@ -108,64 +131,87 @@ export function expressRateLimiter(
         }
 
-        const queryComplexity = getQueryTypeComplexity(queryAST, variables, typeWeightObject);
+        const queryParser = new ASTParser(typeWeightObject, variables);
```
great cleanup

donjewey commented on src/rateLimiters/slidingWindowCounter.ts:
```diff
@@ -153,18 +155,54 @@ class SlidingWindowCounter implements RateLimiter {
+            tokensNeededFromCurrentWindow === 0
+                ? 0
+                : windowSizeSeconds * (tokensNeededFromCurrentWindow / currentTokens);
+
```
excellent work here

donjewey reviewed (approved):
My hat's off to you Evan. This was very thorough, thoughtful, and easy to follow along!
